File: yetutil/src/main/java/net/yet/util/app/DownloadTask.java
package net.yet.util.app;

import android.app.DownloadManager;
import android.app.DownloadManager.Request;
import android.content.Context;
import android.net.Uri;

import net.yet.file.SdAppFile;
import net.yet.util.Util;
import net.yet.util.XMap;
import net.yet.util.log.xlog;

import java.io.File;

public class DownloadTask {
	public static final String FILE = "download_tasks";

	public static final int START = 0;
	public static final int FAILED = 1;
	public static final int SUCCESS = 2;

	public String url;
	public long id = -1;
	public int status = START;
	public boolean actionInstall = false;

	public String fireMsg = null;
	public String filepath = null;

	public DownloadTask() {

	}

	public DownloadTask(String url) {
		this.url = url;
	}

	public void setResult(boolean success) {
		this.status = success ? SUCCESS : FAILED;
		if (success) {
			File file = new File(filepath);
			xlog.INSTANCE.d("下载成功:", filepath);
			if (actionInstall) {
				Util.installApk(file);
			}
		} else {
			Util.toast("下载文件失败!");
		}
		if (fireMsg != null) {
			// MsgCenter.fireN(fireMsg, this);
		}
	}

	/**
	 * @param title
	 *            "更新"
	 * @param extName
	 *            ".apk"
	 */
	public void startAndInstall(String title, String extName) {
		this.actionInstall = true;
		File file = SdAppFile.INSTANCE.tempFile(extName);
		this.filepath = file.getAbsolutePath();

		DownloadManager dm = (DownloadManager) App.get().getSystemService(Context.DOWNLOAD_SERVICE);
		DownloadManager.Request req = new Request(Uri.parse(url));
		req.setTitle(title);
		req.setDestinationUri(Uri.fromFile(file));

		this.id = dm.enqueue(req);

		XMap map = XMap.load(FILE);
		map.put("" + id, this);
		map.save();
	}
}

File: yetutil/src/main/java/yet/util/app/DownloadTask.kt
package yet.util.app

import android.app.DownloadManager.Request
import android.net.Uri
import com.google.gson.JsonObject
import yet.database.MapTable
import yet.file.SdAppFile
import yet.json.getValue
import yet.json.setValue
import yet.util.Util
import yet.util.log.logd
import java.io.File

class DownloadTask(val jo: JsonObject) {

	var id: Long by jo
	var url: String by jo
	var action: String by jo

	var filepath: String by jo


	fun processAction(di: DownloadInfo) {
		val file = File(filepath)
//		val file = File(di.localUri!!.path)
		if (action == ACTION_INSTALLL) {
			Util.installApk(file)
		}
	}


	companion object {
		val downMap = MapTable("downloadTasks")

		val ACTION_INSTALLL = "actionInstall"

		fun onComplete(di: DownloadInfo) {

			val jo = downMap.getJsonObject("${di.id}")
			downMap.remove("${di.id}")
			if (jo != null) {
				val task = DownloadTask(jo)
				if (di.success) {
					task.processAction(di)
				} else {
					logd("下载失败 ", task.url)
				}
			}
		}

		fun downloadAndInstall(url: String, title: String, extName: String, desc: String = "") {
			val task = DownloadTask(JsonObject())
			task.url = url
			task.action = ACTION_INSTALLL

			val file = SdAppFile.tempFile(extName)
			task.filepath = file.absolutePath

			val dm = App.downloadManager
			val req = Request(Uri.parse(url))
			if (title.isNotEmpty()) {
				req.setTitle(title)
			}
			req.setDestinationUri(Uri.fromFile(file))
			if (desc.isNotEmpty()) {
				req.setDescription(desc)
			}
			task.id = dm.enqueue(req)

			downMap.put(task.id.toString(), task.jo.toString())

		}

	}
}